keras_cv/training/contrastive/contrastive_trainer.py

```diff
@@ -0,0 +1,260 @@
+                x,
+                self.augmenters[0](x, training=True),
+                self.augmenters[1](x, training=True),
+                y,
+            )
+        else:
+            return (
+                x,
+                self.augmenters[0](x, training=True),
+                self.augmenters[1](x, training=True),
+            )
+
+    def train_step(self, data):
+        if self.include_probe:
+            if len(data) != 4:
+                raise ValueError(
+                    "Targets must be provided when `include_probe` is True"
+                )
+            images, augmented_images_0, augmented_images_1, labels = data
+        else:
+            if len(data) != 3:
+                raise ValueError(
+                    "Targets must not be provided when `include_probe` is False"
+                )
+            images, augmented_images_0, augmented_images_1 = data
+
+        with tf.GradientTape() as tape:
+            features_0 = self.encoder(augmented_images_0, training=True)
+            features_1 = self.encoder(augmented_images_1, training=True)
+
+            projections_0 = self.projectors[0](features_0, training=True)
+            projections_1 = self.projectors[1](features_1, training=True)
+
+            loss = self.compiled_loss(
+                projections_0, projections_1, regularization_losses=self.encoder.losses
+            )
+
+        gradients = tape.gradient(
+            loss,
+            self.encoder.trainable_weights
+            + self.projectors[0].trainable_weights
+            + self.projectors[1].trainable_weights,
+        )
+
+        self.optimizer.apply_gradients(
+            zip(
+                gradients,
+                self.encoder.trainable_weights
+                + self.projectors[0].trainable_weights
+                + self.projectors[1].trainable_weights,
+            )
+        )
+        self.loss_metric.update_state(loss)
+
+        if self.include_probe:
+            with tf.GradientTape() as tape:
+                features = tf.stop_gradient(self.encoder(images, training=False))
+                class_logits = self.probing_top(features, training=True)
+                probe_loss = self.probe_loss(labels, class_logits)
+            gradients = tape.gradient(probe_loss, self.probing_top.trainable_weights)
+            self.probe_optimizer.apply_gradients(
+                zip(gradients, self.probing_top.trainable_weights)
+            )
+            self.probe_loss_metric.update_state(probe_loss)
+            for metric in self.probe_metrics:
+                metric.update_state(labels, class_logits)
+
+        return {metric.name: metric.result() for metric in self.metrics}
+
+    def call(self, inputs):
+        raise NotImplementedError(
+            "ContrastiveTrainer.call() is not implemented - please call your model directly."
+        )
+
+
+# TODO(ianstenbit): Adapt or share this. Currently shamelessly stolen from https://github.com/keras-team/keras-cv/pull/705
```

Comment:
Lets move this to  keras_cv/utils/training.py as an actual API.

Reply:
Done